File: js/algos/sorting/merge_sort.js
function mergeSort(arr){
	function _merge(v,start,mid,end){
		let i=start;
		let j = mid+1;
		let n = 0;
		let tmp = new Array((end-start)+1);
		while(i <= mid && j <= end){
			if(v[i] <= v[j]){
				tmp[n++] = v[i++];
			}
			else{
				tmp[n++] = v[j++];
			}
		}
		while(i <= mid){
			tmp[n++] = v[i++];
		}
		while(j <= end){
			tmp[n++] = v[j++];
		}
		for(i=start;i<=end;++i){
			v[i] = tmp[i - start];
		}
	}
	function _merge_sort(arr,start,end){
		if(start < end){
			let mid = Math.floor((start+end)/2)
			_merge_sort(arr,start,mid);
			_merge_sort(arr,mid+1,end);
			_merge(arr,start,mid,end);
		}
	}
	_merge_sort(arr,0,arr.length-1);
}
arr = [3,2,1,6,7,9];
mergeSort(arr);
console.log(arr);

File: python/algos/sorting/merge_sort.py
def merge_sort(array):
	def _merge(start,mid,end):
		i = start
		j = mid+1
		n = 0
		tmp = [0] * ((end-start)+1)
		while i <= mid and j <= end:
			if array[i] <= array[j]:
				tmp[n] = array[i]
				i+=1
			else:
				tmp[n] = array[j]
				j+=1

			n+=1

		for i in range(i,mid+1):
			tmp[n] = array[i]
			n+=1
		for j in range(j,end+1):
			tmp[n] = array[j]
			n+=1

		for i in range(start,end+1):
			array[i] = tmp[i - start]

	def _merge_sort(start,end):
		if start < end:
			mid = (start+end)//2
			_merge_sort(start,mid)
			_merge_sort(mid+1,end)
			_merge(start,mid,end)

	_merge_sort(0,len(array)-1)

arr = [3,2,1,6,9,7]
merge_sort(arr)
print(arr)